File: packages/provider-codex/src/failure.ts
import type { ProviderErrorDiagnostics, ProviderFailureFacts } from '@demicodes/core'
import { readHttpFailure, readHttpFailureRecord } from '@demicodes/provider'
import { parseJsonOrString } from '@demicodes/utils'
import { z } from 'zod'

/**
 * A Codex usage limit says when it lifts: `resets_at` in epoch seconds and
 * `resets_in_seconds` from when it was sent. Any other value reads as absent.
 */
const codexLimitSchema = z.looseObject({
  resets_at: z.number().optional().catch(undefined),
  resets_in_seconds: z.number().optional().catch(undefined),
})

/**
 * Where the limit sits in a Codex failure: the error of a stream `error`
 * event, of a WebSocket envelope's `event`, of a `response.failed` event's
 * response, or of an HTTP failure's body.
 */
const codexFailureSchema = z.looseObject({
  error: codexLimitSchema.optional().catch(undefined),
  event: z.looseObject({ error: codexLimitSchema.optional().catch(undefined) })
    .optional()
    .catch(undefined),
  response: z.looseObject({ error: codexLimitSchema.optional().catch(undefined) })
    .optional()
    .catch(undefined),
})

/**
 * Reads a failure record the Codex provider produced
 * (`docs/provider-errors-and-retries.md` § Reading a failure): the usage
 * limit's reset, and otherwise the standard reading of an HTTP failure.
 */
export function readCodexFailure(
  diagnostics: ProviderErrorDiagnostics,
  receivedAt: string
): ProviderFailureFacts {
  const limit = codexLimit(diagnostics)
  if (limit?.resets_at !== undefined)
    return { retryAt: new Date(limit.resets_at * 1000).toISOString() }
  if (limit?.resets_in_seconds !== undefined) {
    const lifts = Date.parse(receivedAt) + limit.resets_in_seconds * 1000
    return { retryAt: new Date(lifts).toISOString() }
  }
  return readHttpFailure(diagnostics, receivedAt)
}

/** The usage limit a failure record carries, when it carries one. */
function codexLimit(
  diagnostics: ProviderErrorDiagnostics
): z.infer<typeof codexLimitSchema> | undefined {
  const text = diagnostics.source === 'http'
    ? readHttpFailureRecord(diagnostics)?.body
    : diagnostics.upstream
  if (!text)
    return undefined
  const failure = codexFailureSchema.safeParse(parseJsonOrString(text))
  if (!failure.success)
    return undefined
  return failure.data.error
    ?? failure.data.event?.error
    ?? failure.data.response?.error
}

File: crates/provider-codex/src/failure.rs
//! How the Codex provider reads its failure records
//! (`failures-and-recovery.md` § Reading a failure): a usage limit says when
//! it lifts, in the failure's own fields, and anything else falls back to
//! the standard reading.

use demi_core::{FailureSource, ProviderErrorDiagnostics, ProviderFailureFacts, Timestamp};
use demi_provider::{HttpFailureRecord, quota::unix_seconds, read_http_failure};

/// The facts of a Codex failure record: `resets_at`, a Unix time in seconds,
/// or `resets_in_seconds`, counted from when the failure was received, of
/// the first error object among a stream event's `error`, a WebSocket
/// envelope's `event.error`, a failed response's `response.error`, or an
/// HTTP failure body's `error`. A value that is not a number reads as
/// absent; without either, the standard `Retry-After` reading applies.
pub fn read_codex_failure(
    diagnostics: &ProviderErrorDiagnostics,
    received_at: Timestamp,
) -> ProviderFailureFacts {
    let text = match diagnostics.source {
        FailureSource::Http => HttpFailureRecord::read(diagnostics).map(|record| record.body),
        _ => diagnostics.upstream.clone(),
    };
    let limit = text
        .and_then(|text| serde_json::from_str::<serde_json::Value>(&text).ok())
        .and_then(|failure| usage_limit(&failure).cloned());
    if let Some(limit) = limit {
        if let Some(resets_at) = limit.get("resets_at").and_then(serde_json::Value::as_f64) {
            return ProviderFailureFacts {
                retry_at: unix_seconds(resets_at),
            };
        }
        if let Some(seconds) = limit
            .get("resets_in_seconds")
            .and_then(serde_json::Value::as_f64)
        {
            let lifts = received_at.as_millisecond() as f64 / 1000.0 + seconds;
            return ProviderFailureFacts {
                retry_at: unix_seconds(lifts),
            };
        }
    }
    read_http_failure(diagnostics, received_at)
}

/// The error object a Codex failure carries its usage limit in.
fn usage_limit(failure: &serde_json::Value) -> Option<&serde_json::Map<String, serde_json::Value>> {
    let nested = |outer: &str| failure.get(outer)?.get("error")?.as_object();
    failure
        .get("error")
        .and_then(serde_json::Value::as_object)
        .or_else(|| nested("event"))
        .or_else(|| nested("response"))
}
